core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -733,7 +735,7 @@ private[spark] class ExecutorAllocationManager(
 
         // If this is the last stage with pending tasks, mark the scheduler queue as empty
         // This is needed in case the stage is aborted for any reason
-        if (stageAttemptToNumTasks.isEmpty && stageAttemptToNumSpeculativeTasks.isEmpty) {
+        if (stageAttemptToNumTasks.isEmpty && stageAttemptToPendingSpeculativeTasks.isEmpty) {
```

Comment:
Should it be:
```suggestion
        if (stageAttemptToNumTasks.isEmpty && stageAttemptToPendingSpeculativeTasks.isEmpty && stageAttemptToSpeculativeTaskIndices.isEmpty) {
```

`stageAttemptToNumSpeculativeTasks` used to include both pending/running speculative tasks but not it's splitted.